beatrix/src/test/java/org/killbill/billing/beatrix/integration/TestCatalogForcedPhaseChangeWithDryRun.java

```diff
@@ -0,0 +1,181 @@
+import org.killbill.billing.beatrix.util.InvoiceChecker.ExpectedInvoiceItemCheck;
+import org.killbill.billing.catalog.DefaultPlanPhasePriceOverride;
+import org.killbill.billing.catalog.api.*;
+import org.killbill.billing.entitlement.api.DefaultEntitlementSpecifier;
+import org.killbill.billing.entitlement.api.Entitlement;
+import org.killbill.billing.entitlement.api.SubscriptionEventType;
+import org.killbill.billing.invoice.api.DryRunArguments;
+import org.killbill.billing.invoice.api.DryRunType;
+import org.killbill.billing.invoice.api.Invoice;
+import org.killbill.billing.invoice.api.InvoiceItemType;
+import org.killbill.billing.payment.api.PluginProperty;
+import org.killbill.billing.platform.api.KillbillConfigSource;
+import org.testng.Assert;
+import org.testng.annotations.Test;
+
+import java.math.BigDecimal;
+import java.util.*;
+
+public class TestCatalogForcedPhaseChangeWithDryRun extends TestIntegrationBase {
+
+    @Override
+    protected KillbillConfigSource getConfigSource(final Map<String, String> extraProperties) {
+        final Map<String, String> allExtraProperties = new HashMap<String, String>(extraProperties);
+        allExtraProperties.put("org.killbill.catalog.uri", "catalogs/testCatalogForcedPhaseChange");
+        return super.getConfigSource(null, allExtraProperties);
+    }
+
+    @Test(groups = "slow")
+    public void testDryRunWithChangePlanInMiddleOfSubscriptionWithPhaseSpecifiedOnNewCatalog() throws Exception {
+        final LocalDate initialDate = new LocalDate(2022, 6, 28);
+        clock.setDay(initialDate);
+
+        final Account account = createAccountWithNonOsgiPaymentMethod(getAccountData(null));
+        final PlanPhaseSpecifier spec = new PlanPhaseSpecifier("pistol-monthly", null);
+        busHandler.pushExpectedEvents(NextEvent.CREATE, NextEvent.BLOCK, NextEvent.INVOICE, NextEvent.INVOICE_PAYMENT, NextEvent.PAYMENT);
+        final UUID createdEntitlementId = entitlementApi.createBaseEntitlement(account.getId(), new DefaultEntitlementSpecifier(spec, null, null, null), UUID.randomUUID().toString(), null, null, false, true, ImmutableList.<PluginProperty>of(), callContext);
+        final Entitlement createdEntitlement = entitlementApi.getEntitlementForId(createdEntitlementId, callContext);
+        assertListenerStatus();
+
+        final LocalDate futureDate = new LocalDate(2022, 7, 13);
+        final DryRunArguments dryRunSubscriptionActionArg = new TestDryRunArguments(DryRunType.SUBSCRIPTION_ACTION,
+                                                                                    "Gun",
+                                                                                    ProductCategory.BASE,
+                                                                                    BillingPeriod.MONTHLY,
+                                                                                    null,
+                                                                                    PhaseType.EVERGREEN,
```

Comment:
Could you do 2 dry-runs, one with null target phase and one with the `EVERGREEN` - that way it becomes clear the code works as expected?

Reply:
Hmmm, I already did 4 tests:
- at start of subscription with the phase specified
- at start of subscription without phase specified **(null target phase)**
- in middle of subscription with phase specified
- in middle of subscription without phase specified

I think it covers your use questions. Or maybe you prefer to have it inside a single unit test?